### PR TITLE
fix(NODE-3852,NODE-3854,NODE-3856): Misc typescript fixes for 4.3.1

### PR DESCRIPTION
**Pull request by baileympearson**
### Description

This PR includes three Typescript fixes / improvements.

#### What is changing?

1. NODE-3856 - allows users to user a document of their collection schema type as a filter in `findX` operations
2. NODE-3854 - update the return types of `findOneAnd<op>` to include the `_id` field.  The behavior of these functions is now aligned with the general `findX` methods
3. NODE-3852  - Adds support for _some_ recursive schemas

##### Is there new documentation needed for these changes?

Probably.  I think we should document exactly what is and is not supported somewhere once we flesh out exactly what we're supporting and we're not supporting.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/types/community/collection/findX.test-d.ts [hunk 1]
```diff
@@ -300,3 +300,223 @@ expectAssignable<SchemaWithUserDefinedId | null>(await schemaWithUserDefinedId.f
...
+ * recursive union types are not supported
+ */
+interface Node {
+  next: Node | null;
+}
+
+declare const nodeCollection: Collection<Node>;
+
+// circular type error is thrown
+// @ts-expect-error
```
Comment (baileympearson) [inline test/types/community/collection/findX.test-d.ts:443]: I talked with Neal about this - `expectError` was unable to catch the recursive type error that's thrown.  `// @ts-expect-error` effectively does the same thing

File: test/types/community/collection/findX.test-d.ts [hunk 2]
```diff
@@ -300,3 +300,223 @@ expectAssignable<SchemaWithUserDefinedId | null>(await schemaWithUserDefinedId.f
...
+
+coll5.findOne({
+  branches: [
+    {
+      id: 'asdf'
+    }
+  ]
+});
+
+// type inference works on non-recursive properties but only at the top level
```
Comment (baileympearson) [inline test/types/community/collection/findX.test-d.ts:511]: I'll fix this comment.  Type inference works on properties whose values are literals and not objects (this is the same as functionality on `main`).

File: test/types/community/collection/findX-recursive-types.test-d.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  title?: string;
+  directories: Directory[];
+}
+
+interface Directory {
+  id: number;
+  name: string;
+  title?: string;
+  branchId: number;
+  files: (number | Directory)[];
```
Comment (baileympearson) [inline test/types/community/collection/findX-recursive-types.test-d.ts:134]: This example closely resembles the example ImRodry posted in his PR here: https://github.com/mongodb/node-mongodb-native/pull/3043

File: src/mongo_types.ts [hunk 4]
```diff
@@ -497,6 +502,16 @@ export type NestedPaths<Type> = Type extends
   : // eslint-disable-next-line @typescript-eslint/ban-types
   Type extends object
   ? {
-      [Key in Extract<keyof Type, string>]: [Key, ...NestedPaths<Type[Key]>];
+      [Key in Extract<keyof Type, string>]: Type[Key] extends Type
```
Comment (dariakp) [inline src/mongo_types.ts:505]: @baileympearson any chance we could annotate these conditions with comments indicating the expected scenario to make it easier to follow?

**Review by dariakp — CHANGES_REQUESTED** [review]
Just a small thing from me, to make this easier to follow (and possibly amend later)

File: src/mongo_types.ts (hunk 4 above)
Comment (baileympearson) [reply src/mongo_types.ts]: Done!

**Review by nbbeeken — APPROVED** [review]
LGTM!

**Review by dariakp — APPROVED** [review]

**Pull request merged**